### ISSUE TITLE
Raster cache hits subject to scaling issues

### ISSUE THREAD

**Issue by tvolkert** [flutter/flutter#17448]
https://github.com/flutter/engine/pull/5180 introduced a regression whereby the screenshotting code produces pixels that are different than those that are actually being rendered on the device or emulator:

![regression](https://user-images.githubusercontent.com/15253456/39840296-c53beda4-5393-11e8-9cbe-e40f0a7e02d4.gif)

This regression surfaced in basically every golden file test, so I'm marking it as a blocker for dev releases.

**Comment by tvolkert** [comment]
Ok, I compared the PNG generated by the [screenshotting code path](https://github.com/flutter/engine/blob/37e20af6a324cd4cfa1324b8c644712ef8224359/shell/common/rasterizer.cc#L162) against the pixels on device (as seen by `adb shell screencap`):

1. They are different
2. This is a regression that was introduced in flutter/engine#5180

Here's the results with flutter/engine#5180 reverted:

![before-5180](https://user-images.githubusercontent.com/15253456/39887825-0774e888-5448-11e8-9196-05934510685e.gif)

And here's the results after flutter/engine#5180

![after-5180](https://user-images.githubusercontent.com/15253456/39887850-1296a4d6-5448-11e8-8097-725ae2599b15.gif)

**Comment by tvolkert** [comment]
Strangely enough, this is fixed if I bypass the raster cache [here](https://github.com/flutter/engine/blob/37e20af6a324cd4cfa1324b8c644712ef8224359/shell/common/rasterizer.cc#L179).

I'm gonna send a PR to do just that to unblock our dev releases, but I'm at a loss as to why bypassing the raster cache fixes this (especially when _not_ bypassing the raster cache _fixed_ #16754).  Further, bypassing the raster cache gets us further away from what we're doing when we render to the display...

**Comment by tvolkert** [comment]
Re-opening, as this isn't fixed, but removing TODAY label as we're no longer blocked.

**Comment by tvolkert** [comment]
Bypassing or not bypassing the raster cache appears to be a red herring in that it tickles the rendering code path and can appear to fix the issue, but while it will fix one screenshot that was broken, it will break another screenshot that wasn't broken...

This is causing huge flakiness in screendiff tests inside Google - re-marking as a release blocker.

**Comment by nsreenath** [comment]
![screenshot_20180515-003007 01](https://user-images.githubusercontent.com/3154529/40017668-b0269200-57d7-11e8-81c6-c90ea3468bc0.png)

The flutter logo in Flutter Gallery 0.4.2 appears pixelated in my Android 7.0 Moto G4 Plus. Is it because of this issue ?

**Comment by tvolkert** [comment]
@nsreenath depends... do you observe different results in the different screenshot types:

```
$ flutter screenshot --type=device
$ flutter screenshot --type=rasterizer --observatory-port=<port>
```

If the former yields a crisp image, and the latter yields the pixelation, then that's this issue.  If not, then it's something else.

**Comment by nsreenath** [comment]
> The flutter logo in Flutter Gallery 0.4.2 appears pixelated in my Android 7.0 Moto G4 Plus. Is it because of this issue ?

@tvolkert The above screenshot is taken from Android's native screenshot. That pixelation is there in the app, not just when taking screenshot.

**Comment by tvolkert** [comment]
Then it's not this issue

**Comment by tvolkert** [comment]
There are two issues here:

1. The `ignore_raster_cache` parameter to [`LayerTree::Preroll()`](https://github.com/flutter/engine/blob/ea9f96b34327382404862127c8f120187fab8c02/flow/layers/layer_tree.h#L27) wasn't in fact causing the raster cache to be ignored.  It was causing us to [skip updating the raster cache result](https://github.com/flutter/engine/blob/ea9f96b34327382404862127c8f120187fab8c02/flow/layers/picture_layer.cc#L18), but we weren't clearing the result either, meaning that when we painted, the leftover cache result could [still report as valid](https://github.com/flutter/engine/blob/ea9f96b34327382404862127c8f120187fab8c02/flow/layers/picture_layer.cc#L36).   This issue is fixed by https://github.com/flutter/engine/pull/5300.

1. When populating the raster cache, we calculate the physical pixels by [applying the appropriate scale](https://github.com/flutter/engine/blob/ea9f96b34327382404862127c8f120187fab8c02/flow/raster_cache.cc#L82), then we [round up](https://github.com/flutter/engine/blob/ea9f96b34327382404862127c8f120187fab8c02/flow/raster_cache.cc#L87=L88) when we allocate the cache image, yet we don't adjust our scale matrix to account for the possibility of partial pixels when we [draw to the canvas](https://github.com/flutter/engine/blob/ea9f96b34327382404862127c8f120187fab8c02/flow/raster_cache.cc#L111).  Then when we later get a raster cache hit and [draw from the cache to the canvas](https://github.com/flutter/engine/blob/ea9f96b34327382404862127c8f120187fab8c02/flow/layers/picture_layer.cc#L39=L45), we scale the _rounded-up image size_ with the _raw scale factor_, exposing us to possible scaling issues.

<span>#</span>2 explains the fuzziness seen in the screendiff tests.  Further:
* The fact that it's triggered by partial pixels explains why for any given test, it was only seen on some display sizes.
* The fact that this was only seen in raster cache hits explains why it was flaky in tests (these were flutter driver tests, so we didn't have fine control over every frame that was rendered).

And finally <span>#</span>1 explains why the problem was still seen when specifying `ignore_raster_cache`.  Further:
* The fact that the `ignore_raster_cache` parameter _did_ keep us from re-populating the raster cache result explains why specifying the parameter _affected_ the test results, even though it didn't fix them.

**Comment by tvolkert** [comment]
FTR, this also explains why the switch to use the main compositor context (flutter/engine#5180) triggered this -- because it introduced the possibility of raster cache hits.  When we were using a fresh compositor context, we only ever had raster cache misses and thus were sidestepping this problem altogether.

**Comment by tvolkert** [comment]
https://github.com/flutter/flutter/issues/17448#issuecomment-390109592 wasn't right - while there does appear to be an issue with our raster cache hit code path that's yielding low-quality images, we do appear to be doing things correctly in the cache scaling in the places I referred to.

After discussing with @abarth and @chinmaygarde, we're going to keep supporting the `ignore_raster_cache` parameter and have the screenshotting code path use it, which will fix the OP issue.  Thus, I'm closing this as fixed, and I filed https://github.com/flutter/flutter/issues/17731 to follow-up with the issue of raster cache hits occasionally yielding low-quality (fuzzy) images.
